### PR TITLE
8303466: C2: failed: malformed control flow. Limit type made precise with MaxL/MinL

### PR DESCRIPTION
**Pull request by eme64**
**Context**

During `PhaseIdealLoop::do_unroll`, we hack the loop-limit, and subtract `stride` from it. We have to prevent underflow on that subtract. Currently, we do this with a `CMoveI`. The problem with this: `CMoveI` is not smart enough to generate a precise type. For example, there are many cases where the input types get better, and underflow is not possible anymore. But the `CMoveI` does not detect this, and still has type `min_jint..hi`.

We have the same issue in `PhaseIdealLoop::adjust_limit`, where we use `CMoveL` to implement long max/min. The types are not as precise as they could and should be.

**Problem**

The imprecise type is used for the zero-trip-guard. It does not fold to false, even though the data-path into the post loop does constant fold to `TOP`. The graph breaks, and assert `malformed control flow` triggers.

Details: In these cases, we have the super-unrolled main-loop (SuperWord'ed, then further unrolled) directly leading to a vectorized post-loop. The effect is that there is no `region/phi` merging main-exit and main-zero-trip-guard. So the types are already more narrow here. It may be possible that the values are such that we find out that we should never enter the vectorized post-loop. But if data finds out and control does not, we get a broken graph.
Note: we have pre-loop. Then a main-loop and vectorized post loop. Then we merge the main-zero-trip-guard. And at the end we have the scalar post loop.

I have already recently fixed a bug around this `CMoveI`. https://github.com/openjdk/jdk/commit/5a4945c0d95423d0ab07762c915e9cb4d3c66abb I would now like to have a more satisfactory fix, that properly propagates the types.

**Solution**

`PhaseIdealLoop::adjust_limit` already converts the limit from int to long, and does all computations in long, including taking max/min with a `CMoveL`. I now use the so far unused `MaxL/MinL`. I implemented some missing `Value/Identity` components for it. Since `MaxL/MinL` is not implemented in the backend, I just expand it in macro-expansion to a `CMoveL`. At that point the loop-opts are over, and it is most likely ok that we do not make the types more precise after this.

I take the same approach for `PhaseIdealLoop::do_unroll`: convert limits to long, do subtraction in long, take `MinL/MaxL` to clamp it to the int-range (prevent subtraction underflow).

**Discussion**

This solution seems much cleaner to me, and I hope that we will see less bugs because of imprecise types in the limit computation, which were often due to the `CMove` not being smart enough to analyze all inputs (it would have to recognize a multitude of patterns, for the Cmp inputs and the direct inputs to the CMove - we currently do not do that, but just take the union of the input types - this is very inprecise).

There is a bit of an overhead here: We use longs even though we only want to have int values. But I think we should prefer a clean implementation here, with correct type computation. The performance impact is probably non-existent on 64-bit machines anyway.

**Caveat**

I found some cases with the same assert `malformed control flow` that are most likely skeleton/assertion predicate bugs [JDK-8288981](https://bugs.openjdk.org/browse/JDK-8288981). Some of those cases were new patterns, for example where we PreMainPost a main loop.

I hope that this fix here at least reduces the frequency of failures significantly.

**Testing**

I added 2 regression tests. Our fuzzer seems to spit out examples regularly, so that gives us extra coverage.

Tested up to `tier5` and stress testing. Performance testing **running...**

**Future Work**

We should implement `MaxL/MinL` in the backend. We should also use them during parsing. This would also allow to `SuperWord` the instruction, on the platforms that support it.

Should we add such an assert during IGVN? I think after IGVN, we should never have a `MultiBranchNode` that does not have the required number of outputs, right? We could add it to `VerifyIterativeGVN`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303466](https://bugs.openjdk.org/browse/JDK-8303466): C2: failed: malformed control flow. Limit type made precise with MaxL/MinL

### Reviewers
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**) ⚠️ Review applies to [fdf6b08a](https://git.openjdk.org/jdk/pull/13269/files/fdf6b08a50a0b90e0df496046c984a24d965cf28)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [33e1ad54](https://git.openjdk.org/jdk/pull/13269/files/33e1ad54d0322bdbc50c85fbea6f1ada0963f3ef)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [4a13b00b](https://git.openjdk.org/jdk/pull/13269/files/4a13b00b1d38f082e2632cf391b117ad7cff05d1)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [ec052b17](https://git.openjdk.org/jdk/pull/13269/files/ec052b1761777434d972b88c23d47b1fe3f74175)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13269/head:pull/13269` \
`$ git checkout pull/13269`

Update a local copy of the PR: \
`$ git checkout pull/13269` \
`$ git pull https://git.openjdk.org/jdk.git pull/13269/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13269`

View PR using the GUI difftool: \
`$ git pr show -t 13269`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13269.diff">https://git.openjdk.org/jdk/pull/13269.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13269#issuecomment-1497303625)

### CODE REVIEW

**Comment by eme64** [comment]
@rwestrel 
> Is the PhaseIdealLoop::adjust_limit() change required or is it some cleanup?

At first I only fixed `PhaseIdealLoop::do_unroll`. That fixed my regression test examles. But, once that fix was introduced, another test failed, and that was because now the type of the `CMove` in `PhaseIdealLoop::adjust_limit` was not precise enough. So it seemed the best solution to fix them together, since they both have issues with `CMove`, and change the `limit`.

> Have you run performance testing to be safe?

I'll do that, will report back.

**Review by rwestrel — APPROVED** [review]
Looks good to me.

**Comment by eme64** [comment]
@rwestrel thanks for the review!
The performance testing looks ok, I cannot see a significant runtime change.

File: src/hotspot/share/opto/loopTransform.cpp [hunk 1]
```diff
@@ -2288,74 +2288,32 @@ void PhaseIdealLoop::do_unroll(IdealLoopTree *loop, Node_List &old_new, bool adj
...
-      if ((stride_con > 0 && (java_subtract(limit_type->_lo, stride_con) < limit_type->_lo)) ||
-          (stride_con < 0 && (java_subtract(limit_type->_hi, stride_con) > limit_type->_hi))) {
-        // No underflow.
-        new_limit = new SubINode(limit, stride);
+      // Limit is not constant. Int subtraction could lead to underflow.
+      // (1) Convert to long.
+      Node* limit_l = new ConvI2LNode(limit);
+      register_new_node(limit_l, get_ctrl(limit));
+      Node* stride_l = new ConvI2LNode(stride);
+      register_new_node(stride_l, get_ctrl(limit));
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopTransform.cpp:2296]: Can we make Long constant (`_igvn.longcon(stride)`) instead since `stride` is constant? Similar to `underflow_clamp_l`. My concern is you set control to constant which is not `Root`.

**Review by vnkozlov — COMMENTED** [review]
If we have several unrolls we will have chain of `MaxL/MinL` nodes. Will the chain be folded by IGVN?

File: src/hotspot/share/opto/addnode.cpp [hunk 2]
```diff
@@ -1258,6 +1258,48 @@ const Type *MinINode::add_ring( const Type *t0, const Type *t1 ) const {
   return TypeInt::make( MIN2(r0->_lo,r1->_lo), MIN2(r0->_hi,r1->_hi), MAX2(r0->_widen,r1->_widen) );
 }
 
+const Type* MaxLNode::add_ring(const Type* t0, const Type* t1) const {
+  const TypeLong* r0 = t0->is_long();
+  const TypeLong* r1 = t1->is_long();
+
+  return TypeLong::make( MAX2(r0->_lo,r1->_lo), MAX2(r0->_hi,r1->_hi), MAX2(r0->_widen,r1->_widen) );
```
Comment (chhagedorn) [inline src/hotspot/share/opto/addnode.cpp:1265]:
```suggestion
  return TypeLong::make(MAX2(r0->_lo, r1->_lo), MAX2(r0->_hi, r1->_hi), MAX2(r0->_widen, r1->_widen));
```

File: src/hotspot/share/opto/addnode.cpp [hunk 3]
```diff
@@ -1258,6 +1258,48 @@ const Type *MinINode::add_ring( const Type *t0, const Type *t1 ) const {
...
+  }
+
+  return MaxNode::Identity(phase);
+}
+
+const Type* MinLNode::add_ring(const Type* t0, const Type* t1) const {
+  const TypeLong* r0 = t0->is_long();
+  const TypeLong* r1 = t1->is_long();
+
+  return TypeLong::make( MIN2(r0->_lo,r1->_lo), MIN2(r0->_hi,r1->_hi), MIN2(r0->_widen,r1->_widen) );
```
Comment (chhagedorn) [inline src/hotspot/share/opto/addnode.cpp:1286]:
```suggestion
  return TypeLong::make(MIN2(r0->_lo, r1->_lo), MIN2(r0->_hi, r1->_hi), MIN2(r0->_widen, r1->_widen));
```

File: src/hotspot/share/opto/addnode.hpp [hunk 4]
```diff
@@ -322,28 +322,36 @@ class MinINode : public MaxNode {
 // MAXimum of 2 longs.
 class MaxLNode : public MaxNode {
 public:
-  MaxLNode(Node *in1, Node *in2) : MaxNode(in1, in2) {}
+  MaxLNode(Compile* C, Node *in1, Node *in2) : MaxNode(in1, in2) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/addnode.hpp:325]:
```suggestion
  MaxLNode(Compile* C, Node* in1, Node* in2) : MaxNode(in1, in2) {
```

File: src/hotspot/share/opto/addnode.hpp [hunk 5]
```diff
@@ -322,28 +322,36 @@ class MinINode : public MaxNode {
...
+    init_flags(Flag_is_macro);
+    C->add_macro_node(this);
+  }
   virtual int Opcode() const;
-  virtual const Type *add_ring(const Type*, const Type*) const { return TypeLong::LONG; }
-  virtual const Type *add_id() const { return TypeLong::make(min_jlong); }
-  virtual const Type *bottom_type() const { return TypeLong::LONG; }
+  virtual const Type* add_ring(const Type*, const Type*) const;
```
Comment (chhagedorn) [inline src/hotspot/share/opto/addnode.hpp:330]:
```suggestion
  virtual const Type* add_ring(const Type* t0, const Type* t1) const;
```

File: src/hotspot/share/opto/addnode.hpp [hunk 6]
```diff
@@ -322,28 +322,36 @@ class MinINode : public MaxNode {
...
   int min_opcode() const { return Op_MinL; }
+  virtual Node* Identity(PhaseGVN* phase);
 };
 
 //------------------------------MinLNode---------------------------------------
 // MINimum of 2 longs.
 class MinLNode : public MaxNode {
 public:
-  MinLNode(Node *in1, Node *in2) : MaxNode(in1, in2) {}
+  MinLNode(Compile* C, Node *in1, Node *in2) : MaxNode(in1, in2) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/addnode.hpp:343]:
```suggestion
  MinLNode(Compile* C, Node* in1, Node* in2) : MaxNode(in1, in2) {
```

File: src/hotspot/share/opto/addnode.hpp [hunk 7]
```diff
@@ -322,28 +322,36 @@ class MinINode : public MaxNode {
...
+    init_flags(Flag_is_macro);
+    C->add_macro_node(this);
+  }
   virtual int Opcode() const;
-  virtual const Type *add_ring(const Type*, const Type*) const { return TypeLong::LONG; }
-  virtual const Type *add_id() const { return TypeLong::make(max_jlong); }
-  virtual const Type *bottom_type() const { return TypeLong::LONG; }
+  virtual const Type* add_ring(const Type*, const Type*) const;
```
Comment (chhagedorn) [inline src/hotspot/share/opto/addnode.hpp:348]:
```suggestion
  virtual const Type* add_ring(const Type* t0, const Type* t1) const;
```

**Review by chhagedorn — APPROVED** [review]
That looks good to me and indeed much cleaner! I only have some minor code style comments.

> There is a bit of an overhead here: We use longs even though we only want to have int values. But I think we should prefer a clean implementation here, with correct type computation. The performance impact is probably non-existent on 64-bit machines anyway.

I agree with that.

> I have already recently fixed a bug around this CMoveI. https://github.com/openjdk/jdk/commit/5a4945c0d95423d0ab07762c915e9cb4d3c66abb I would now like to have a more satisfactory fix, that properly propagates the types.

I've had a feeling that we are revisiting this again at some point.

> Should we add such an assert during IGVN? I think after IGVN, we should never have a MultiBranchNode that does not have the required number of outputs, right? We could add it to VerifyIterativeGVN.

That would be a good idea to investigate in an RFE - maybe also for other nodes to assert on well-known input/output patterns. We've had such problems before after CCP with `If` nodes with only one out projection.

**Comment by eme64** [comment]
I added the idea about verifying out-proj of `MultiBranch` to this RFE [JDK-8298951](https://bugs.openjdk.org/browse/JDK-8298951).

File: src/hotspot/share/opto/loopTransform.cpp (hunk 1 above)
Comment (eme64) [reply src/hotspot/share/opto/loopTransform.cpp:2296]: Good point, will replace it with constant. Yes, I had the ctrl wrong, it should be root.

File: src/hotspot/share/opto/addnode.cpp [hunk 8]
```diff
@@ -1258,6 +1258,48 @@ const Type *MinINode::add_ring( const Type *t0, const Type *t1 ) const {
...
+  const TypeLong* r1 = t1->is_long();
+
+  return TypeLong::make(MAX2(r0->_lo, r1->_lo), MAX2(r0->_hi, r1->_hi), MAX2(r0->_widen, r1->_widen));
+}
+
+Node* MaxLNode::Identity(PhaseGVN* phase) {
+  const TypeLong* t1 = phase->type(in(1))->is_long();
+  const TypeLong* t2 = phase->type(in(2))->is_long();
+
+  // Can we determine minimum statically?
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/addnode.cpp:1272]:
```suggestion
  // Can we determine maximum statically?
```

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by vnkozlov** [comment]
> I think in many cases, the type does not underflow, and the `MaxL/MinL` can be removed completely.

I would disagree. In many cases `limit` is variable. At least I want you to check generated code for such case and add it to your test. Even if additional latency of several CMov (to which you convert Max/Min nodes) vs one node is negligible the bigger size of generated code may affect inlining.

> TLDR: @vnkozlov is it ok if I investingate & test `MaxL/MinL` and `ConvI2L / ConvL2I` folding in a follow-up RFE?

Depending on result of investigation of generated code.

**Comment by jaskarth** [comment]
> However, if that does not work, I think it now also fails to remove the repeated ConvI2L / ConvL2I. We would have to add more IGVN optimizations to fold things more.

I think you're running into an issue where some nodes created by counted loop expansion aren't properly passed onto the IGVN worklist- I found the same thing while trying to investigate some strange code generation from small loops. If you make that follow-up RFE I would be happy to attach the cases that I found as well.

**Comment by eme64** [comment]
@jaskarth please send me those cases, if it is many then maybe better via email. I'm generally working on doing verification of that kind, see [JDK-8298951](https://bugs.openjdk.org/browse/JDK-8298951).

**Comment by jaskarth** [comment]
I only had a handful of cases so I've attached them in [this gist](https://gist.github.com/jaskarth/878648eecaf74e168d5499c5961b4715).
I found these a few weeks ago, but looking back I think I have misremembered the problem chain. Examples 3-5 may be a different bug as they deal with `ConvL2I->ConvI2L` chains instead of `ConvI2L->ConvL2I` chains as you are seeing, as the latter has an Identity() transform defined while it seems the former does not- I apologize for the noise if the issues are unrelated. Examples 1 and 2 could perhaps still be useful in diagnosing the issue, as they describe cases where ideal transforms that do exist aren't taken. I tried looking into that bug myself a while ago but didn't get far.

JDK-8298951 is exciting, it'll make reasoning about middle-end optimizations a lot easier :)

**Comment by vnkozlov** [comment]
> But I think I can do the same with just collapsing `SubL -> MaxL -> SubL -> MaxL` to `SubL -> MaxL`. That may be cleaner.

I prefer this if you can do it. So you have sequence (after folding `Conv` nodes) 
```
MaxL(SubL(MaxL(SubL(limit, stride), min_int), stride*2), min_int);
```

Yes, I think it can be collapsed to:
```
MaxL(SubL(limit, stride*3), min_int);
```

If in any point of chain `limit` become `min_int` it will stay `min_int`  (even if `stride` is `max_int`) because you use Long arithmetic and we have "small" limit on unrolling (16?).
If it does not hit min_int the result it similar to SubL(SubL((limit, stride), stride*2).
So you just need to correctly collect `stride*N` values.

**Comment by eme64** [comment]
@vnkozlov I added some more IGVN optimizations that help to fold the `SubL -> MaxL` chains.

1. `fold_subI_no_underflow_pattern` in `MaxLNode::Ideal`. Collapses `SubL -> MaxL->SubL -> MaxL` to a simple `SubL -> MaxL`.  

2. `ConvI2LNode::Identity` can now convert `I2L(L2I(x))` => `x`. We need this, so that the Casts are not in the way for the first optimization.

I added verification, that these optimizations are really taken:
https://github.com/openjdk/jdk/blob/33e1ad54d0322bdbc50c85fbea6f1ada0963f3ef/test/hotspot/jtreg/compiler/loopopts/TestLoopLimitSubtractionsCollapse.java#L50-L68

File: src/hotspot/share/opto/addnode.cpp [hunk 9]
```diff
@@ -1258,6 +1258,136 @@ const Type *MinINode::add_ring( const Type *t0, const Type *t1 ) const {
...
+    return in(2);
+  }
+
+  return MaxNode::Identity(phase);
+}
+
+// Collapse the "addition with overflow-protection" pattern, and the symetrical
+// "subtraction with underflow-protection" pattern. These are created during the
+// unrolling, when we have to adjust the limit by subtracting the stride, but want
+// to protect agains underflow: MaxL(SubL(limit, stride), min_jint).
```
Comment (vnkozlov) [inline src/hotspot/share/opto/addnode.cpp:1285]: May add note that SubL node is replaced with AddL and reversed stride ( I assume that is what happened here).

File: src/hotspot/share/opto/addnode.cpp [hunk 10]
```diff
@@ -1258,6 +1258,136 @@ const Type *MinINode::add_ring( const Type *t0, const Type *t1 ) const {
...
+//
+//   x  con1  con2
+//   |    |    |
+//   |   AddLNode (new_con)
+//   |    |
+//  AddLNode  clamp1
+//        |    |
+//       Max/MinL (n)
+//
+Node* fold_subI_no_underflow_pattern(Node* n, PhaseGVN* phase) {
```
Comment (vnkozlov) [inline src/hotspot/share/opto/addnode.cpp:1308]: Move this method and it comment before `MaxLNode::add_ring` so all MaxL and MinL method stay together.

**Review by vnkozlov — APPROVED** [review]
Looks good. Just minor comments.

**Review by TobiHartmann — APPROVED** [review]
Still looks good to me.

File: src/hotspot/share/opto/addnode.hpp [hunk 11]
```diff
@@ -322,28 +322,38 @@ class MinINode : public MaxNode {
...
-  virtual const Type *add_id() const { return TypeLong::make(max_jlong); }
-  virtual const Type *bottom_type() const { return TypeLong::LONG; }
+  virtual const Type* add_ring(const Type* t0, const Type* t1) const;
+  virtual const Type* add_id() const { return TypeLong::make(max_jlong); }
+  virtual const Type* bottom_type() const { return TypeLong::LONG; }
   virtual uint ideal_reg() const { return Op_RegL; }
   int max_opcode() const { return Op_MaxL; }
   int min_opcode() const { return Op_MinL; }
+  virtual Node* Identity(PhaseGVN* phase);
+  virtual Node* Ideal(PhaseGVN *phase, bool can_reshape);
```
Comment (chhagedorn) [inline src/hotspot/share/opto/addnode.hpp:356]:
```suggestion
  virtual Node* Ideal(PhaseGVN* phase, bool can_reshape);
```

**Review by chhagedorn — APPROVED** [review]
Updates look good!

**Comment by eme64** [comment]
Thanks @chhagedorn @TobiHartmann @vnkozlov @rwestrel for the reviews and suggestions!
The assert will now still fail with the fuzzer occasionally because of the `assertion / skeleton predicate` bug that @chhagedorn is already working on for a while. But I hope this fix will drastically reduce the rate of fuzzer failures with this assert.
/integrate
